Homework 23

OneAndOnlyAdam opened the pull request.

OneAndOnlyAdam committed:
Homework19

Had some issues with the last step but got it figured out finally.

OneAndOnlyAdam committed:
Homework23

Help!  I ran it to see if things are working and it gave this message in IntelliJ:
BUILD FAILED in 3s
5 actionable tasks: 5 executed

When I click on the report link I keep getting these errors:
org.gradle.api.internal.tasks.testing.TestSuiteExecutionException: Could not complete execution for Gradle Test Executor 3.
The number at the end goes up by one every time I run it.

OneAndOnlyAdam committed:
Homework23 attempt2

Without the try/catch in BaseTest for:

public String getDeletedPlaylistMsg() {
        notificationMsg = wait.until(ExpectedConditions.visibilityOfElementLocated(By.cssSelector("div.success.show")));
        return notificationMsg.getText();

I get the following error in the report:
java.lang.NullPointerException: Cannot invoke "org.openqa.selenium.support.ui.WebDriverWait.until(java.util.function.Function)" because "this.wait" is null

With the try/catch in BaseTest for:

public String getDeletedPlaylistMsg() {
    try {
        notificationMsg = wait.until(ExpectedConditions.visibilityOfElementLocated(By.cssSelector("div.success.show")));
    } catch (NullPointerException npe) {
        //return notificationMsg.getText();
    }
    return notificationMsg.getText();

I get the following error in the report:

java.lang.NullPointerException: Cannot invoke "org.openqa.selenium.WebElement.getText()" because "this.notificationMsg" is null

Berichicko commented on src/test/java/Homework23.java:
```diff
@@ -0,0 +1,23 @@
+
+    //Prerequisite - empty playlist to delete
+
+    @Test
+    public void deletePlaylist() throws InterruptedException{
+        String expectedPLayListDeletedMessage = "Deleted playlist \"Golden Girls.\"";
+
+        provideEmail("adam.johnson@testpro.io");
+        providePassword("1Te$t$tudent");
+        clickSubmit();
+        openPlaylist();
+        clickDeletePlaylistBtn();
+        Assert.assertEquals(getDeletedPlaylistMsg(), expectedPLayListDeletedMessage);
+    }
```
**Use in test methods, the following logical split:   GIVEN – WHEN - THEN**
   
   For example
   
   
![Clean_Test](https://github.com/Test-Pro-Bootcamp/A51-QA-Automation-Framework/assets/104761482/a282b186-cc4f-4c5d-a61f-df14f94611a7)

Berichicko commented:
**HINT**

1. **BasePage and BaseTest** 

**BaseTest** — Implements the necessary methods for work with webdriver.

The BaseTest class contains common setup and teardown methods, as well as any shared functionality or configurations required by multiple test classes.

**BaseTest**

The problem we are trying to solve here is the reuse of common behaviors across test classes avoiding coding duplication and centralizing these actions at one point.

**@BeforeSuite**	- The annotated method will be run before all tests in this suite have run.

**@BeforeMethod** -	The annotated method will be run before all the test methods in the current class have been run.

**@AfterMethod**	- The annotated method will be run after each test method.

_Example without a Base Test Class_

![image](https://user-images.githubusercontent.com/104761482/233972627-2d7ffdc7-6a2f-4e9f-8a3a-a1b4a7a47bc4.png)


**BasePage**  — class serves as a foundation or parent class for individual page classes. It contains common methods and functionalities that are shared across multiple pages in an application.

_Example without a Base Page Class_

![image](https://user-images.githubusercontent.com/104761482/236891493-7f7013c7-fde2-4bf9-9084-3aa7483df8e3.png)


2. **Test in Page Object Model (POM)**

![Clean_Test](https://user-images.githubusercontent.com/104761482/236863813-f30d7bfc-e89d-4419-bdea-4b1517155e39.png)

3.  **Page Object Model and Page Factory**

![POM_vs_Factory_1](https://user-images.githubusercontent.com/104761482/236864030-34b89652-460d-4440-aa91-894332e90654.png)

![POM_vs_Factory_2](https://user-images.githubusercontent.com/104761482/236864098-d3ffe636-7ba9-484e-a72f-8b03ff024659.png)

4.  Please, use **Encapsulation**

Encapsulation is one of the four fundamental OOP concepts. The other three are inheritance, polymorphism, and abstraction.

You should use encapsulation of non-static and non-final fields.

If a variable is used only by other methods of that class, it should be declared as private.

Make all the locators and WebElements private, because private keyword in java allows most restrictive access to variables and methods and offer strongest form of Encapsulation. 

For example

    ```
    private By shuffleBtnLocator = By.cssSelector(".btn-shuffle-all");
    private By firstSongLocator = By.cssSelector(".play");

```
   @FindBy(css = "img.avatar")
    private WebElement avatar
```


Make the variables protected, because protected access gives the subclass a chance to use the helper method or variable, while preventing a nonrelated class from trying to use it.

```
    protected Actions actions;
    protected WebDriverWait wait;
    protected WebDriver driver;
```

6. Try to analyse status tests

_For example_

![Test status_en](https://user-images.githubusercontent.com/104761482/236867158-08437918-56e2-462b-b7c9-5629d9f16e74.png)

![Test failed_en](https://user-images.githubusercontent.com/104761482/236867187-6613ad47-d9a9-4d85-a1cf-3ca1ecf8a347.png)


![Test error_en](https://user-images.githubusercontent.com/104761482/236867225-898c86e0-89b6-4eb1-a707-1de6cae472d4.png)

__________________________________________________________________________________________________________________________

OneAndOnlyAdam committed:
Homework23...

I am at a loss... It looks to me like the playlist isn't getting deleted but the error is for not getting the deleted playlist message.

java.lang.AssertionError: expected [Deleted playlist "Golden Girls."] but found []
	at org.testng.Assert.fail(Assert.java:110)

Berichicko commented:
You got this problem, please correct it. Please finish your homework.

![image](https://github.com/Test-Pro-Bootcamp/A51-QA-Automation-Framework/assets/104761482/ff4c66e5-027f-401a-878e-c961a727c96f)

**HINT**

**Step 1**

![image](https://github.com/Test-Pro-Bootcamp/A51-QA-Automation-Framework/assets/104761482/967cfd9a-c1a9-4b9b-9d55-b7a2155d34b7)


```
package PageObjectModel;

import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.interactions.Actions;
import org.openqa.selenium.support.PageFactory;
import org.openqa.selenium.support.ui.ExpectedConditions;
import org.openqa.selenium.support.ui.WebDriverWait;

import java.time.Duration;

public class BasePage {

    protected WebDriver driver;
    protected WebDriverWait wait;
    protected Actions actions;

    // Constructor for BasePage class
    public BasePage(WebDriver givenDriver){
        driver = givenDriver;
        wait = new WebDriverWait(driver, Duration.ofSeconds(3));
        actions = new Actions(driver);
        PageFactory.initElements(driver, this);
    }

    // Method to click on a web element
    protected void click(WebElement webElement){
        wait.until(ExpectedConditions.elementToBeClickable(webElement)).click();
    }

    // Method to find a web element
    protected WebElement findElement(WebElement webElement){
        return wait.until(ExpectedConditions.visibilityOf(webElement));
    }

}
```

**Step 2**

![image](https://github.com/Test-Pro-Bootcamp/A51-QA-Automation-Framework/assets/104761482/794203ea-eb96-4953-b4ff-a091f9eb4a13)


```
package PageObjectModel;

import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.support.FindBy;

public class HomePage extends BasePage {
    @FindBy(css = ".playlist:nth-child(3)")
    private WebElement emptyPlaylistLocator;

    @FindBy(css = ".btn-delete-playlist")
    private WebElement deletePlaylistLocator;

    @FindBy(css = "div.success.show")
    private WebElement notificationMsgLocator;

    @FindBy(css = "your-avatar-selector")
    private WebElement avatarLocator;

    // Constructor for HomePage class
    public HomePage(WebDriver givenDriver) {
        super(givenDriver);
    }

    // Method to open the playlist
    public void openPlaylist() {
       findElement(emptyPlaylistLocator);
        click(emptyPlaylistLocator);
    }

    // Method to click on the delete playlist button
    public void clickDeletePlaylistBtn()  {
        findElement(deletePlaylistLocator);
        click(deletePlaylistLocator);
    }

    // Method to get the text of the notification message
    public String getDeletedPlaylistMsg() {
        findElement(notificationMsgLocator);
        return notificationMsgLocator.getText();
    }

    // Method to check if the avatar selector is displayed
    public boolean isAvatarDisplayed() {
       findElement(avatarLocator);
        return avatarLocator.isDisplayed();
    }
}
```

**Step 3**

![image](https://github.com/Test-Pro-Bootcamp/A51-QA-Automation-Framework/assets/104761482/34a50c29-c392-4963-9b6a-9e2c3ae7cd89)


```
import io.github.bonigarcia.wdm.WebDriverManager;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.chrome.ChromeDriver;
import org.openqa.selenium.chrome.ChromeOptions;
import org.testng.annotations.AfterMethod;
import org.testng.annotations.BeforeMethod;
import org.testng.annotations.BeforeSuite;
import org.testng.annotations.Parameters;

import java.time.Duration;

public class BaseTest {

    public WebDriver driver;

    // Setup WebDriverManager for ChromeDriver
    @BeforeSuite
    static void setupClass() {
        WebDriverManager.chromedriver().setup();
    }

    // Launch Chrome browser and navigate to the BaseUrl
    @BeforeMethod
    @Parameters({"BaseUrl"})
    public void launchBrowser(String BaseUrl) throws InterruptedException {
        ChromeOptions options = new ChromeOptions();
        options.addArguments("--disable-notifications","--remote-allow-origins=*", "--incognito","--start-maximized");
        options.setExperimentalOption("excludeSwitches", new String[]{"enable-automation"});
        driver = new ChromeDriver(options);
        driver.manage().timeouts().implicitlyWait(Duration.ofSeconds(5));
        driver.get(BaseUrl);
    }

    // Close the browser after each test method
    @AfterMethod
    public void closeBrowser() {
        driver.quit();
    }
}
```

**Step 4**

![image](https://github.com/Test-Pro-Bootcamp/A51-QA-Automation-Framework/assets/104761482/fa7aeed8-aaed-41ff-801a-66236b1fe899)

**RESULT**

![image](https://github.com/Test-Pro-Bootcamp/A51-QA-Automation-Framework/assets/104761482/c88c6c64-edab-49ad-9852-4e3ae14fc133)

OneAndOnlyAdam committed:
Homework23....again

Got it working.  FINALLY

Thanks for the help!!!

Berichicko commented:
Hello,

I have checked your homework 23, and I wanted to let you know that you're doing great! Keep up the excellent work!
Best wishes from the HW team (Alex, [aliaksandr@testpro.io](mailto:aliaksandr@testpro.io))!

![804deb64dc406a20a75da05d91a1a69e](https://user-images.githubusercontent.com/104761482/235961479-9afb2841-b657-46c9-99f8-2402f933b843.jpg)

![image](https://github.com/Test-Pro-Bootcamp/A51-QA-Automation-Framework/assets/104761482/41023b32-fd31-4615-80a4-bfabf25193ef)
